dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/apikey/ApiTokenServiceImpl.java

```diff
@@ -0,0 +1,151 @@
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.security.apikey;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.nio.charset.StandardCharsets;
+import java.security.SecureRandom;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.lang3.StringUtils;
+import org.hisp.dhis.user.User;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import com.google.common.base.Preconditions;
+import com.google.common.hash.Hashing;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Service
+@Transactional
+public class ApiTokenServiceImpl implements ApiTokenService
+{
+    private static final Long DEFAULT_EXPIRE_TIME_IN_MILLIS = TimeUnit.DAYS.toMillis( 30 );
+
+    private final ApiTokenStore apiTokenStore;
+
+    public ApiTokenServiceImpl( ApiTokenStore apiTokenStore )
+    {
+        checkNotNull( apiTokenStore );
+
+        this.apiTokenStore = apiTokenStore;
+    }
+
+    @Override
+    public List<ApiToken> getAll()
+    {
+        return this.apiTokenStore.getAll();
+    }
+
+    @Override
+    public List<ApiToken> getAllOwning( User currentUser )
+    {
+        return apiTokenStore.getAllOwning( currentUser );
+    }
+
+    @Override
+    @Transactional( readOnly = true )
+    public ApiToken getWithKey( String key, User currentUser )
+    {
+        return apiTokenStore.getByKey( key, currentUser );
+    }
+
+    @Override
+    public ApiToken getWithKey( String key )
+    {
+        return apiTokenStore.getByKey( key );
+    }
+
+    @Override
+    @Transactional
+    public void save( ApiToken apiToken )
+    {
+        throw new IllegalArgumentException(
+            "Tokens can not be saved, all tokens must be created with the createToken() method." );
+
+    }
+
+    @Override
+    @Transactional
+    public void update( ApiToken apiToken )
+    {
+        checkNotNull( apiToken, "Token can not be null" );
+        Preconditions.checkArgument( StringUtils.isNotEmpty( apiToken.getKey() ), "Token key can not be null" );
+        checkNotNull( apiToken.getCreatedBy(), "Token must have an owner" );
+        checkNotNull( apiToken.getExpire(), "Token must have an expire value" );
+        checkNotNull( apiToken.getType(), "Token must have an type value" );
+        checkNotNull( apiToken.getVersion(), "Token must have an version value" );
+
+        apiTokenStore.update( apiToken );
+
+        // Invalidate cache here or let cache expire ?
+    }
+
+    @Override
+    @Transactional
+    public void delete( ApiToken apiToken )
+    {
+        apiTokenStore.delete( apiToken );
+        // Invalidate cache here or let cache expire ?
```

Comment:
I think we should invalidate the cache here in case the api token is being deleted for security reasons.